### PR TITLE
[SPARK-41228][SQL] Rename & Improve error message for `COLUMN_NOT_IN_GROUP_BY_CLAUSE`.

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to rename `COLUMN_NOT_IN_GROUP_BY_CLAUSE` to `MISSING_AGGREGATION`.

Also, improve its error message.

### Why are the changes needed?

The current error class name and its error message doesn't illustrate the error cause and resolution correctly.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

```
./build/sbt “sql/testOnly org.apache.spark.sql.SQLQueryTestSuite*”
```

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -785,6 +779,13 @@
       "Malformed Protobuf messages are detected in message deserialization. Parse Mode: <failFastMode>. To process malformed protobuf message as null result, try setting the option 'mode' as 'PERMISSIVE'."
     ]
   },
+  "MISSING_AGGREGATION" : {
+    "message" : [
+      "The non-aggregating expression <expression> is based on columns which are not participating in the GROUP BY clause.",
+      "Add the columns or the expression to the GROUP BY, aggregate the expression, or use \"any_value(<expression>)\" if you do not care which of the values within a group is returned."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:785]: `expression` is quoted already, correct? How this `\"any_value(<expression>)\"` looks like `"any_value("...")"`?

Comment (itholic) [reply core/src/main/resources/error/error-classes.json:785]: Yeah, it looks like the example below:

```
[MISSING_AGGREGATION] The non-aggregating expression "c2" is based on columns which are not participating in the GROUP BY clause.
Add the columns or the expression to the GROUP BY, aggregate the expression, or use "any_value("c2")" if you do not care which of the values within a group is returned.;
```

Can we fix it to ``` `any_value("...")` ``` or something ?

Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:785]: How about just leave `expression` in the error format:
```
... or use <expression> if you do not care which ...
```
and pass the message parameter:
```scala
    new AnalysisException(
      errorClass = "MISSING_AGGREGATION",
      messageParameters = Map("expression" -> toSQLExpr(new AnyValue(expression)))
```

Comment (itholic) [reply core/src/main/resources/error/error-classes.json:785]: Cool. Let me update soon.

Comment (itholic) [reply core/src/main/resources/error/error-classes.json:785]: I just tried your suggestion, and I checked the first <expression> one is also affected, so it looks a bit wired IMO.
For example, the error message will be shown like:

```
... The non-aggregating expression "anyvalue(c2)" is based on columns ...
... aggregate the expression, or use "any_value(c2)" if you do not care ...
```

Maybe should we separate the parameters into 2 parts like <expression> and <expression_any_value> ??

Comment (itholic) [reply core/src/main/resources/error/error-classes.json:785]: Or maybe we simply recommend to use the `any_value()` function as below:

```
... aggregate the expression, or use `any_value()` for <expression> if you do not care ...
```

something like that?

**Comment by itholic** [comment]
Fixed the Python test first.

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:785]: > "anyvalue(c2)" .. "any_value(c2)"

hmm, could you open a PR and override `prettyName` in `AnyValue`, and pass `prettyName` to `FirstLast.validateIgnoreNullExpr()`. Also, cc @vitaliili-db

**Comment by MaxGekk** [comment]
@itholic Please, rebase on the recent master.

**Comment by MaxGekk** [comment]
@itholic Could you fix the test failure. It seems it is related to your changes:
```
python/pyspark/sql/tests/pandas/test_pandas_udf_grouped_agg.py.test_invalid_args
"COLUMN_NOT_IN_GROUP_BY_CLAUSE" does not match "[MISSING_AGGREGATION] The non-aggregating expression "v" is based on columns which are not participating in the GROUP BY clause.
Add the columns or the expression to the GROUP BY, aggregate the expression, or use "any_value("v")" if you do not care which of the values within a group is returned.;
Aggregate [id#668L], [id#668L, plus_one(v#674)#687 AS plus_one(v)#688]
```

File: python/pyspark/sql/tests/pandas/test_pandas_udf_grouped_agg.py [hunk 2]
```diff
@@ -475,7 +475,7 @@ def test_invalid_args(self):
         mean_udf = self.pandas_agg_mean_udf
 
         with QuietTest(self.sc):
-            with self.assertRaisesRegex(AnalysisException, "nor.*aggregate function"):
+            with self.assertRaisesRegex(AnalysisException, "[MISSING_AGGREGATION]"):
```
Comment (itholic) [inline python/pyspark/sql/tests/pandas/test_pandas_udf_grouped_agg.py:478]: We should test this with similar logic with `checkError` in Scala side.

Let me create the JIRA and update soon after getting some feedback from community.

File: core/src/main/resources/error/error-classes.json (hunk 1 above)
Comment (MaxGekk) [reply core/src/main/resources/error/error-classes.json:830]: @itholic Since the PR has been merged, I am waiting for when you update your PR.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@itholic Please, construct `AnyValue`.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @itholic.

**Pull request closed without merging**